src/opm/output/eclipse/AggregateWellData.cpp

```diff
@@ -0,0 +1,773 @@
+                    sWell[1 - 1] = swprop(M::liquid_surface_rate, pp.OilRate);
+                }
+
+                if (pp.hasProductionControl(PP::WRAT)) {
+                    sWell[2 - 1] = swprop(M::liquid_surface_rate, pp.WaterRate);
+                }
+
+                if (pp.hasProductionControl(PP::GRAT)) {
+                    sWell[3 - 1] = swprop(M::gas_surface_rate, pp.GasRate);
+                }
+
+                if (pp.hasProductionControl(PP::LRAT)) {
+                    sWell[4 - 1] = swprop(M::liquid_surface_rate, pp.LiquidRate);
+                }
+
+                if (pp.hasProductionControl(PP::RESV)) {
+                    sWell[5 - 1] = swprop(M::rate, pp.ResVRate);
+                }
+
+                if (pp.hasProductionControl(PP::THP)) {
+                    sWell[6 - 1] = swprop(M::pressure, pp.THPLimit);
+                }
+
+                if (pp.hasProductionControl(PP::BHP)) {
+                    sWell[7 - 1] = swprop(M::pressure, pp.BHPLimit);
+                }
+            }
+            else if (well.isInjector(sim_step)) {
+                const auto& ip = well.getInjectionProperties(sim_step);
+
+                using IP = ::Opm::WellInjector::ControlModeEnum;
+
+                if (ip.hasInjectionControl(IP::THP)) {
+                    sWell[6 - 1] = swprop(M::pressure, ip.THPLimit);
+                }
+
+                if (ip.hasInjectionControl(IP::BHP)) {
+                    sWell[7 - 1] = swprop(M::pressure, ip.BHPLimit);
+                }
+            }
+
+            sWell[10 - 1] = swprop(M::length, datumDepth(well, sim_step));
+        }
+    } // SWell
+
+    namespace XWell {
+        std::size_t entriesPerWell(const std::vector<int>& inteHead)
+        {
+            // INTEHEAD(27) = NXWELZ
+            assert ((inteHead[27 - 1] > 123) &&
+                    "XWEL must allocate at least 124 elements per well");
+
+            return inteHead[27 - 1];
+        }
+
+        Opm::RestartIO::Helpers::WindowedArray<double>
+        allocate(const std::vector<int>& inteHead)
+        {
+            using WV = Opm::RestartIO::Helpers::WindowedArray<double>;
+
+            return WV {
+                WV::NumWindows{ numWells(inteHead) },
+                WV::WindowSize{ entriesPerWell(inteHead) }
+            };
+        }
+
+        template <class XWellArray>
+        void staticContrib(const ::Opm::Well&     well,
+                           const Opm::UnitSystem& units,
+                           const std::size_t      sim_step,
+                           XWellArray&            xWell)
+        {
+            using M = ::Opm::UnitSystem::measure;
+
+            // xWell[41] = BHP target.
+
+            xWell[41] = well.isInjector(sim_step)
+                ? well.getInjectionProperties (sim_step).BHPLimit
+                : well.getProductionProperties(sim_step).BHPLimit;
+
+            xWell[41] = units.from_si(M::pressure, xWell[41]);
```

Comment:
All xWell assignments are done with plain zero-based indices, whereas the iWell and sWell was done using "one-based   - 1" notation. Any reason for the difference? Will be important to keep straight when using enum constants instead of literals (as requested by @joakim-hove, I agree with the request).